### PR TITLE
kernel: add rtl881au-mw package

### PR DESCRIPTION
**Pull request by plntyk**
another Realtek rtl8812au driver modified by morrownr
used in Arch Linux AUR

scanning works with qemu x86_64 + TP-Link T4U stock

- joining open networks fails
- joining encrypted networks fails

Signed-off-by: Dirk Neukirchen <plntyk.lede@plntyk.name>

### CODE REVIEW

**Comment by plntyk** [comment]
the rtl8812au-ct driver is based on 4.3.14 driver code - this is a modified 5.9.3.2
"morrownr" variant was "recommended" by someone in IRC trying to get it compiled for OpenWrt

its working a little bit better than the
rtl8812au-ct with my usb stick (TP-Link Archer T4U cannot scan more than once)

but has major issues in OpenWrt (not connecting)

There are many different rtl8812au on github but these 2 -gb and -mw seem like candidates that are currently "working" for some ppl on "normal" linux OSes

the aircrack variants seem popular too - havent tested them yet

File: package/kernel/rtl8812au-mw/Makefile [hunk 1]
```diff
@@ -0,0 +1,59 @@
+include $(TOPDIR)/rules.mk
```
Comment (adschm) [inline package/kernel/rtl8812au-mw/Makefile:1]: Please add an SPDX license header for every new file you add.

**Comment by adschm** [comment]
Please remove [RFC] prefix from commit title and duplicate Signed-off-by from commit message.

**Comment by obsy** [comment]
Wait, are you trying to add non working driver to OpeWrt? Why?

**Comment by plntyk** [comment]
> Wait, are you trying to add non working driver to OpeWrt? Why?

done more testing with this + -gb variant :

the driver does work on commandline but not with luci

aircrack-ng variant: https://github.com/aircrack-ng/rtl8812au uses different 5.6.4.2 codebase

changes to make aircrack variant compile might also be small - if anyone is more interested in that leave a comment
I havent looked in detail through the changes between trees

**Comment by musashino205** [comment]
Is the package name in the commit title ```rtl881au-mw``` correct?

**Comment by plntyk** [comment]
> Is the package name in the commit title `rtl881au-mw` correct?

yes but feel free to suggest different title - its some compromise between "original" realtek source thats: 8812au or variants of that and adding prefix + suffixes: "rtl .... -ct" to packagename / module name

maybe module name should be modified - it now clashes with the -ct variant so parallel installation isnt possible

feel free to test & further improve

**Comment by musashino205** [comment]
Ah, it means ```rtl881au-mw``` vs. ```rtl8812au-mw```

**Comment by plntyk** [comment]
sry -force pushed update to fix missing 2

**Comment by plntyk** [comment]
@M0E-lnx feel free to test this or the other package

**Comment by lucky930** [comment]
DKMS make.log for rtl88x2bu-git for kernel 5.10.0-kali8-amd64 (x86_64)
Friday 11 June 2021 12:00:03 PM IST
/bin/sh: 1: bc: not found
make ARCH=x86_64 CROSS_COMPILE= -C /lib/modules/5.10.0-kali8-amd64/build M=/var/lib/dkms/rtl88x2bu/git/build  modules
make[1]: Entering directory '/usr/src/linux-headers-5.10.0-kali8-amd64'
Makefile:1: /usr/src/linux-headers-5.10.0-kali8-common/Makefile: No such file or directory
make[1]: *** No rule to make target '/usr/src/linux-headers-5.10.0-kali8-common/Makefile'.  Stop.
make[1]: Leaving directory '/usr/src/linux-headers-5.10.0-kali8-amd64'
make: *** [Makefile:2378: modules] Error 2


Anyone help on this.??????

**Comment by lucky930** [comment]
$ sudo apt-get install -y linux-headers-$(uname -r) build-essential dkms git libelf-dev bc

running above command helped me to resolve it and then follow below link to install the drivers,

https://github.com/RinCat/RTL88x2BU-Linux-Driver

**Comment by 1715173329** [comment]
https://github.com/immortalwrt/immortalwrt/tree/master/package/kernel/rtl8812au-ac

aircrack variant has some minor issues but works okay on the whole.

**Comment by castillofrancodamian** [comment]
> > Wait, are you trying to add non working driver to OpeWrt? Why?
> 
> done more testing with this + -gb variant :
> 
> the driver does work on commandline but not with luci
> 
> aircrack-ng variant: https://github.com/aircrack-ng/rtl8812au uses different 5.6.4.2 codebase
> 
> changes to make aircrack variant compile might also be small - if anyone is more interested in that leave a comment
> I havent looked in detail through the changes between trees

If the aircrack version is better, I'd like it to be as well.

**Comment by plntyk** [comment]
> If the aircrack version is better, I wish it was in OpenWrt as well.

https://github.com/openwrt/openwrt/pull/4273

I personally do not care which one is picked / selected - aircrack version does not work with luci "scan"

there are several configure switches and code differences between the branches and no sane kernel developer is touching this Realtek code anyway since its a hot mess of a common codebase for all realtek devices where parts depending on "vendor" contract are removed / unavailable (see Makefile and code structure between 8812au, 88x2bu for example)

**Comment by 981213** [comment]
Could this go https://github.com/openwrt/packages instead?
It doesn't quite make sense to have an out-of-tree driver in OpenWrt core when it's not required by any supported device.
Same with #4273

**Comment by cristian-ciobanu** [comment]
A new better driver was released for Realtek 8812AU chipset
https://github.com/morrownr/8812au-20210629

**Comment by castillofrancodamian** [comment]
> A new better driver was released for Realtek 8812AU chipset https://github.com/morrownr/8812au-20210629

But is it compatible with OpenWrt?

**Comment by jpsollie** [comment]
> > A new better driver was released for Realtek 8812AU chipset https://github.com/morrownr/8812au-20210629
> 
> But is it compatible with OpenWrt?

to my experience -- I tried the aircrack drivers - if you patch the makefile a bit and make sure the preprocessor doesn't make any wrong assumption due to linux_version_code != mac80211 backported API, it is pretty compatible.  we could even extend the makefile  (though I didn't test it):
```
CONFIG_RTL8188E = n
CONFIG_RTL8812A = y
CONFIG_RTL8821A = n
CONFIG_RTL8192E = n
CONFIG_RTL8723B = n
CONFIG_RTL8814A = n
CONFIG_RTL8723C = n
CONFIG_RTL8188F = n
CONFIG_RTL8188GTV = n
CONFIG_RTL8822B = n
CONFIG_RTL8723D = n
CONFIG_RTL8821C = n
CONFIG_RTL8710B = n
CONFIG_RTL8192F = n
CONFIG_RTL8822C = n
CONFIG_RTL8814B = n
CONFIG_RTL8723F = n
```
if these are removed from the makefile, and the wanted chipset is passed as a parameter for make, it should maybe be possible to create a menu with wanted chipsets instead, whereby make is called for each chip with a different 'y' value, thus producing another kernel module ...

**Comment by castillofrancodamian** [comment]
> 

So there would be drivers for different chipsets from a single driver?

**Comment by jpsollie** [comment]
> > 
> 
> So there would be drivers for different chipsets from a single driver?

I know it sounds weird, but yes, that's what I meant: from the same codebase you generate kernel modules for different devices.

**Review by jpsollie — APPROVED** [review]

**Comment by cristian-ciobanu** [comment]
Is this pull request ready to merged or it still needs changes or to be reviewed ?
Also can it be updated to use the new source for the driver from here https://github.com/morrownr/8812au-20210629 ?

**Comment by castillofrancodamian** [comment]
I doubt they will merge it as it is a driver that doesn't work properly in OpenWrt.

**Comment by xruifan** [comment]
> > > A new better driver was released for Realtek 8812AU chipset https://github.com/morrownr/8812au-20210629
> > 
> > 
> > But is it compatible with OpenWrt?
> 
> to my experience -- I tried the aircrack drivers - if you patch the makefile a bit and make sure the preprocessor doesn't make any wrong assumption due to linux_version_code != mac80211 backported API, it is pretty compatible. we could even extend the makefile (though I didn't test it):
> 
> ```
> CONFIG_RTL8188E = n
> CONFIG_RTL8812A = y
> CONFIG_RTL8821A = n
> CONFIG_RTL8192E = n
> CONFIG_RTL8723B = n
> CONFIG_RTL8814A = n
> CONFIG_RTL8723C = n
> CONFIG_RTL8188F = n
> CONFIG_RTL8188GTV = n
> CONFIG_RTL8822B = n
> CONFIG_RTL8723D = n
> CONFIG_RTL8821C = n
> CONFIG_RTL8710B = n
> CONFIG_RTL8192F = n
> CONFIG_RTL8822C = n
> CONFIG_RTL8814B = n
> CONFIG_RTL8723F = n
> ```
> 
> if these are removed from the makefile, and the wanted chipset is passed as a parameter for make, it should maybe be possible to create a menu with wanted chipsets instead, whereby make is called for each chip with a different 'y' value, thus producing another kernel module ...

https://github.com/openwrt/openwrt/pull/4273
https://github.com/openwrt/openwrt/pull/4837
By any chance have you tried the updated aircrack-ng variant?

**Comment by 981213** [comment]
I think it's not a good idea to take garbage realtek wifi drivers into OpenWrt. In my experience these drivers don't have a proper mac80211 interface implemented (that's why luci can't work with it properly) and they are unstable. There are USB wifi dongles with upstream linux driver, which will give users a way better experience.
I'm closing this PR now.

**Pull request closed without merging**